core/src/main/java/com/yahoo/oak/HashChunk.java

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+class HashChunk<K, V> extends BasicChunk<K, V> {
+    // HashChunk takes a number of least significant bits from the full key hash
+    // to provide as an index in the EntryHashSet
+    private UnionCodec hashIndexCodec; // to be given
+
+    /*-------------- Members --------------*/
+    private final EntryHashSet<K, V> entryHashSet;
+
+    /*-------------- Constructors --------------*/
+    /**
+     * This constructor is only used when creating the first ever chunk/chunks (without a creator).
+     * The caller might set the creator before returning the HashChunk to the user.
+     *
+     * @param maxItems  is the size of the entries array (not all the entries are going to be in use)
+     *                  IMPORTANT!: it is better to be a power of two,
+     *                  if not the rest of the entries are going to be waisted
+     * @param hashIndexCodec the codec initiated with the right amount of the least significant bits,
+     *                       to be used to get index from the key hash
+     */
+    HashChunk(int maxItems, AtomicInteger externalSize, MemoryManager vMM, MemoryManager kMM,
+        OakComparator<K> comparator, OakSerializer<K> keySerializer,
+        OakSerializer<V> valueSerializer, UnionCodec hashIndexCodec) {
+
+        super(maxItems, externalSize, comparator);
+        assert Math.pow( 2, hashIndexCodec.getFirstBitSize() ) <= maxItems ;
+        assert hashIndexCodec != null;
+
+        this.hashIndexCodec = hashIndexCodec;
+        this.entryHashSet = // must be called after setSecondLevelBitsThreshold
+            new EntryHashSet<>(vMM, kMM, getMaxItems(), keySerializer, valueSerializer,
+                comparator);
+    }
+
+    /**
+     * Create a child HashChunk where this HashChunk object as its creator.
+     */
+    HashChunk<K, V> createChild(UnionCodec hashIndexCodec) {
+        HashChunk<K, V> child =
+            new HashChunk<>(getMaxItems(), externalSize,
+                entryHashSet.valuesMemoryManager, entryHashSet.keysMemoryManager,
+                comparator, entryHashSet.keySerializer, entryHashSet.valueSerializer,
+                hashIndexCodec);
+        updateBasicChild(child);
+        return child;
+    }
+
+    private int calculateKeyHash(K key) {
+        int hashKey = key.hashCode(); // hash can be positive, zero or negative
+        return Math.abs(hashKey); // EntryHashSet doesn't accept negative hashes
```

Comment:
Why not? It has 32 bits.

Reply:
This is because UnionCodec doesn't deal with negative numbers